src/components/lib/sl/sl.c

```diff
@@ -563,29 +595,43 @@ sl_init(void)
 void
 sl_sched_loop(void)
 {
+	struct sl_global *g = sl__globals();
+
 	while (1) {
 		int pending;
 
 		do {
 			thdid_t        tid;
 			int            blocked, rcvd;
 			cycles_t       cycles;
-			tcap_time_t    timeout = 0, thd_timeout;
-			struct sl_thd *t;
+			tcap_time_t    timeout = g->timeout_next, thd_timeout;
+			struct sl_thd *t, *tn;
 
 			/*
 			 * a child scheduler may receive both scheduling notifications (block/unblock
 			 * states of it's child threads) and normal notifications (mainly activations from
 			 * it's parent scheduler).
 			 */
-			pending = cos_sched_rcv(sl__globals()->sched_rcv, RCV_ALL_PENDING, timeout,
+			pending = cos_sched_rcv(g->sched_rcv, RCV_ALL_PENDING, timeout,
 						&rcvd, &tid, &blocked, &cycles, &thd_timeout);
-			if (!tid) continue;
+			if (!tid) goto pending_events;
 
 			t = sl_thd_lkup(tid);
 			assert(t);
-			/* don't report the idle thread */
-			if (unlikely(t == sl__globals()->idle_thd)) continue;
+			/* don't report the idle thread or a freed thread */
+			if (unlikely(t == g->idle_thd || t->state == SL_THD_FREE)) goto pending_events;
+
+			/*
+			 * Failure to take the CS because another thread is holding it and switching to
+			 * that thread cannot succeed because scheduler has pending events causes the event
+			 * just received to be dropped.
+			 * To avoid dropping events, add the events to the scheduler event list and processing all
+			 * the pending events after the scheduler can successfully take the lock.
+			 */
+			sl_thd_event_enqueue(t, blocked, cycles, thd_timeout);
+
+pending_events:
```

Comment:
I don't buy that using `goto`s for this is the best option, but I'm not going to hold up this PR anymore.